cmd/internal/find/primary/meta_test.go

```diff
@@ -1,43 +1,103 @@
 package primary
 
 import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
 	"testing"
 	"time"
 
 	"github.com/puppetlabs/wash/cmd/internal/find/types"
 	"github.com/puppetlabs/wash/cmd/internal/find/params"
+	"github.com/puppetlabs/wash/cmd/internal/find/parser/parsertest"
 	"github.com/stretchr/testify/suite"
 )
 
 type MetaPrimaryTestSuite struct {
-	suite.Suite
+	parsertest.Suite
+	e types.Entry
 }
 
-func (suite *MetaPrimaryTestSuite) SetupTest() {
+func (s *MetaPrimaryTestSuite) SetupTest() {
 	params.StartTime = time.Now()
 }
 
-func (suite *MetaPrimaryTestSuite) TeardownTest() {
+func (s *MetaPrimaryTestSuite) TeardownTest() {
 	params.StartTime = time.Time{}
 }
 
-func (suite *MetaPrimaryTestSuite) TestMetaPrimaryErrors() {
-	_, _, err := metaPrimary.parse([]string{"-m", "-true"})
-	suite.Regexp("-m: key sequences must begin with a '.'", err)
+func (s *MetaPrimaryTestSuite) TestMetaPrimaryErrors() {
+	s.RunTestCases(
+		s.NPETC("-m", "expected a key sequence", false),
+		s.NPETC("-m foo", "key sequences must begin with a '.'", false),
+		s.NPETC("-m .", "expected a key sequence after '.'", false),
+		s.NPETC("-m .[", "expected a key sequence after '.'", false),
+		s.NPETC("-m .key", "expected a predicate expression", false),
+		s.NPETC("-m .key +{", "expected.*closing.*}", false),
+		s.NPETC("-m .key]", `expected an opening '\['`, false),
+		s.NPETC("-m .key[", `expected a closing '\]'`, false),
+	)
+}
 
-	_, _, err = metaPrimary.parse([]string{"-m", "."})
-	suite.Regexp("-m: expected a key sequence after '.'", err)
+func (s *MetaPrimaryTestSuite) TestMetaPrimaryValidInputTruePredicates() {
+	s.RunTestCases(
+		s.NPTC("-m .architecture x86_64 -primary", "-primary", s.e),
+		s.NPTC("-m .blockDeviceMappings[?] .deviceName /dev/sda1 -primary", "-primary", s.e),
+		s.NPTC("-m .cpuOptions.coreCount 4 -primary", "-primary", s.e),
+		s.NPTC("-m .tags[?] .key termination_date -a .value +1h -primary", "-primary", s.e),
+		s.NPTC("-m .tags[?] .key foo -o .key department -primary", "-primary", s.e),
```

Comment:
As a follow-up, I'd like to extend ObjectPredicate/ArrayPredicate to also support the `wash find` operators. That would reduce this test-case to
```
-m .tags[?] .key \( foo -o department \) -primary
```

(Parens are necessary to distinguish ObjectPredicate/ArrayPredicate expressions from the top-level meta primary expression).

Reply:
May also be worth introducing some simple regex support for `meta` keys, because typing out e.g. `blockDeviceMappings` can be annoying, especially when you have to do it more than once. I'm thinking of something like `blockDev*[?] .deviceName /dev/sda1 -primary` where `blockDev*` means "key matching blockDev*".

Reply:
With all the parens stuff in here, it might also not be a bad idea to have the expression parser support something like `\(foo` where it'd strip the parens off. Then the `department` test case could further be shortened to:
```
-m .tags[?] .key \(foo -o department\) -primary
```

Reply:
Those are interesting. I'm not against spending a little time on them, but probably don't want to spend a lot at the moment since you can reasonable accomplish your goal.

Reply:
How does parsing distinguish between `-m .key foo -o .key bar` and `-m .key foo -o -m .key bar`? In other words, how does it know `-o` is part of the meta expression rather than after the meta expression?

P.S. If I read the answer earlier I may have just forgotten it and now can't find it.

Reply:
Don't think it should take too long, maybe a day.

And to answer your question, meta's expression parser stops if it encounters an unknown token. For your examples, `-m .key foo -o .key bar` will be parsed as `-m .key ( foo -o .key bar)`. Here, the predicate's on `meta["key"]` and it returns true if that value is the string "foo" OR if that value is an object with its "key" key set to the string "bar" (if that makes sense). The distinguishing factor there's the `KeySequence Expression` part in the grammar, where `.key` is the KeySequence and `foo -o .key bar` is the expression.

The second example  `-m .key foo -o -m .key bar` should be parsed by the top-level parser as `Meta(.key, foo) -o Meta(.key, bar)` where the `meta` primary will return an `UnknownTokenError` when it encounters the second `-m` token so that it hands it off to the top-level expression parser. However, thanks to https://github.com/puppetlabs/wash/blob/master/cmd/internal/find/parser/expression/parser.go#L184-L195, it currently does not do this -- it returns a `-m: no expression after -o` error so it is a bug. Will fix that.